File: js/trie.js
(function() {
    function Trie() {
        this.root = [null, {}];
    }

    Trie.prototype.add = function( key, value ) {
        var node, ch, curr_node = this.root;
        for(var i=0,k_len=key.length;i<k_len;i++) {
            ch = key.charAt(i);
            node = curr_node[1];
            if(ch in node) {
                curr_node = node[ch];
            } else {
                curr_node = node[ch] = [null,{}];
            }
        }
        curr_node[0] = value;
    };

    Trie.prototype.find = function( key ) {
        var ch, curr_node = this.root;
        for(var i=0,k_len=key.length; i<k_len; i++) {
            ch = key.charAt(i);
            if(ch in curr_node[1]) {
                curr_node = curr_node[1][ch];
            } else {
                return;
            }
        }
        return curr_node[0];
    };

    Trie.prototype.find_prefix = function( key ) {
        var ch, curr_node = this.root;
        var remainder = key;
        for(var i=0,k_len=key.length;i<k_len;i++) {
            ch = key.charAt(i);
            if(ch in curr_node[1]) {
                curr_node = curr_node[1][ch];
            } else {
                return [curr_node[0], remainder];
            }
            remainder = remainder.slice(1,remainder.length);
        }
        return [curr_node[0], remainder];
    };

    function _find_prefix_match( trie, key ) {
        var ch, curr_node = trie.root;
        var remainder = key;
        for(var i=0,k_len=key.length;i<k_len; i++) {
            ch = key.charAt(i);
            if(ch in curr_node[1]) {
                curr_node = curr_node[1][ch];
            } else {
                break;
            }
            remainder = remainder.slice(1,remainder.length);
        }
        return [curr_node,remainder];
    }

    Trie.prototype.find_prefix_matches = function( key ) {
        var l = _find_prefix_match( this, key );
        if(l[1].length>0) {
            return [];
        }
        var d, stack = [l[0]], ret = [];
        while(stack.length>0) {
            d = stack.pop();
            if(d[0]) {
                ret.unshift(d[0]);
            }
            for(var c in d[1]) {
                if(d[1].hasOwnProperty(c)) {
                    stack.push(d[1][c]);
                }
            }
        }
        return ret;
    };
    window['Trie'] = Trie;
}());

File: python/trie.py
# -*- coding: utf-8 -*-
'''
The trie object is based on James Tauber's trie.py
http://jtauber.com/2005/02/trie.py
'''

_SENTINEL = ()

class Trie(object) :
    __slots__ = ['root']

    def __init__( self ) :
        self.root = [None,{}]

    def __getstate__( self ) :
        if any(self.root) :
            return self.root
        else :
            return False

    def __setstate__(self, s) :
        self.root = s

    def __contains__( self, s ) :
        if self.find_full_match(s,_SENTINEL) is _SENTINEL :
            return False
        return True

    def add( self, key, value ) :
        curr_node = self.root
        for ch in key :
            node = curr_node[1]
            if ch in node :
                curr_node = node[ch]
            else :
                curr_node = node[ch] = [None,{}]
        curr_node[0] = value

    def _find_prefix_match( self, key ) :
        curr_node = self.root
        remainder = key
        for ch in key :
            if ch in curr_node[1] :
                curr_node = curr_node[1][ch]
            else :
                break
            remainder = remainder[1:]
        return [curr_node,remainder]

    def find_full_match( self, key, fallback=None ) :
        '''
        Returns the value associated with the key if found else, returns fallback
        '''
        r = self._find_prefix_match( key )
        if not r[1] and r[0] :
            return r[0][0]
        return fallback

    def find_prefix_matches( self, prefix ) :
        l = self._find_prefix_match( prefix )
        if l[1] :
            return []
        stack = [l[0]]
        ret = []
        while stack :
            d = stack.pop()
            if d[0] :
                ret.insert(0,d[0])
            for c in d[1] :
                stack.append(d[1][c])
        return ret
